gc/base/SparseVirtualMemory.cpp

```diff
@@ -0,0 +1,233 @@
+ *******************************************************************************/
+
+#include <string.h>
+#if defined(OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION)
+#include <sys/mman.h>
+#include <sys/errno.h>
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION */
+
+#include "omrcomp.h"
+#include "omrport.h"
+#include "omr.h"
+
+#include "EnvironmentBase.hpp"
+#include "Forge.hpp"
+#include "GCExtensionsBase.hpp"
+#include "Math.hpp"
+#include "Heap.hpp"
+#include "HeapRegionManager.hpp"
+#include "ModronAssertions.h"
+#include "SparseVirtualMemory.hpp"
+#include "SparseAddressOrderedFixedSizeDataPool.hpp"
+
+/****************************************
+ * Initialization
+ ****************************************
+ */
+
+MM_SparseVirtualMemory*
+MM_SparseVirtualMemory::newInstance(MM_EnvironmentBase* env, uint32_t memoryCategory, MM_Heap *in_heap)
+{
+	MM_SparseVirtualMemory* vmem = NULL;
+	vmem = (MM_SparseVirtualMemory*)env->getForge()->allocate(sizeof(MM_SparseVirtualMemory), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+
+	if (vmem) {
+		new (vmem) MM_SparseVirtualMemory(env, in_heap->getPageSize(), in_heap);
+		if (!vmem->initialize(env, memoryCategory)) {
+			vmem->kill(env);
+			vmem = NULL;
+		}
+	}
+
+	return vmem;
+}
+
+bool
+MM_SparseVirtualMemory::initialize(MM_EnvironmentBase* env, uint32_t memoryCategory)
+{
+	uintptr_t in_heap_size = (uintptr_t)_heap->getHeapTop() - (uintptr_t)_heap->getHeapBase();
+	uintptr_t maxHeapSize = _heap->getMaximumMemorySize();
+	uintptr_t regionSize = _heap->getHeapRegionManager()->getRegionSize();
+	uintptr_t regionCount = in_heap_size / regionSize;
+
+	/* Ideally this should be ceilLog2, however this is the best alternative as ceilLog2 currently does not exist */
+	uintptr_t ceilLog2 = MM_Math::floorLog2(regionCount) + 1;
+
+	uintptr_t off_heap_size = (uintptr_t)((ceilLog2 * in_heap_size) / 2);
+	bool success = MM_VirtualMemory::initialize(env, off_heap_size, NULL, NULL, 0, memoryCategory);
+
+	if (success) {
+		void *sparseHeapBase = getHeapBase();
+		_sparseDataPool = MM_SparseAddressOrderedFixedSizeDataPool::newInstance(env, sparseHeapBase, off_heap_size);
+		if ((NULL == _sparseDataPool) || omrthread_monitor_init_with_name(&_largeObjectVirtualMemoryMutex, 0, "SparseVirtualMemory::_largeObjectVirtualMemoryMutex")) {
+			success = false;
+		}
+	}
+
+	if (success) {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_success((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	} else {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_failure((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	}
```

Comment:
minor, but it could really be just one trace point with an extra string param (mapped from success bool)

Reply:
Ok 👍 The only other difference in these traces is that in `j9mm.tdf`, I have labelled `Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_failure` as a `TraceException` and `Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_success` as a `TraceEvent`. If with this new info the one trace statement is still preferred, I can definitely make this change.

Reply:
Honestly, don't what difference it makes. Possibly for tools like HC to filter and show only Exception!?
Do you know @dmitripivkine ?

Reply:
I am not sure we do have customers to use this. May be HC as you said. Do we have different default level for `TraceException`?  The main difference to have two or one tracepoint is possibility to enable/disable them individually. So using two different tracepoints make sense, at least for me. And "failed" case tracepoint might be enabled by default

Reply:
ok, let's leave them separate